File: android/src/main/kotlin/com/changteng/agora_rtc/AgoraRtcEngineActions.kt
package com.changteng.agora_rtc

import android.content.Context
import io.agora.rtc2.ChannelMediaOptions
import io.agora.rtc2.ClientRoleOptions
import io.agora.rtc2.Constants
import io.agora.rtc2.RtcEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

class AgoraRtcEngineActions(
  private val engineHolder: AgoraRtcEngineHolder,
  private val emitter: AgoraEventEmitter
) {
  private val eventHandler = AgoraRtcEngineEventHandler(emitter)
  private val mediaActions = AgoraRtcEngineMediaActions(engineHolder)
  private val recordingActions = AgoraRtcEngineRecordingActions(engineHolder, emitter)

  fun clearRuntimeState() {
    recordingActions.clear()
  }

  fun handleCreateEngine(call: MethodCall, result: Result, applicationContext: Context?) {
    if (engineHolder.rtcEngine != null) {
      result.error("ENGINE_EXISTS", "RtcEngine 已初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val appId = args["appId"] as? String
    if (appId.isNullOrBlank()) {
      result.error("INVALID_ARGS", "appId 为空", null)
      return
    }
    val context = applicationContext
    if (context == null) {
      result.error("NO_CONTEXT", "applicationContext 为空", null)
      return
    }
    try {
      engineHolder.rtcEngine = RtcEngine.create(context, appId, eventHandler)
      engineHolder.bindAllVideoIfReady()
      recordingActions.clear()
      result.success(0)
    } catch (ex: Exception) {
      result.error("ENGINE_CREATE_FAILED", ex.message, null)
    }
  }

  fun handleDestroyEngine(result: Result) {
    if (engineHolder.rtcEngine != null) {
      RtcEngine.destroy()
      engineHolder.clear()
    }
    recordingActions.clear()
    result.success(0)
  }

  fun handleJoinChannel(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val token = args["token"] as? String ?: ""
    val channelId = args["channelId"] as? String
    if (channelId.isNullOrBlank()) {
      result.error("INVALID_ARGS", "channelId 为空", null)
      return
    }
    val uid = (args["uid"] as? Number)?.toInt() ?: 0
    val optionsMap = args["options"] as? Map<*, *>
    val options = buildChannelMediaOptions(optionsMap)
    val code = engine.joinChannel(token, channelId, uid, options)
    result.success(code)
  }

  fun handleLeaveChannel(result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val code = engine.leaveChannel()
    result.success(code)
  }

  fun handleUpdateChannelMediaOptions(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val optionsMap = args["options"] as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "options 为空", null)
      return
    }
    val options = buildChannelMediaOptions(optionsMap)
    val code = engine.updateChannelMediaOptions(options)
    result.success(code)
  }

  fun handleSetChannelProfile(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val profile = (args["profile"] as? Number)?.toInt()
    if (profile == null) {
      result.error("INVALID_ARGS", "profile 为空", null)
      return
    }
    val code = engine.setChannelProfile(profile)
    result.success(code)
  }

  fun handleRenewToken(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val token = args["token"] as? String
    if (token.isNullOrEmpty()) {
      result.error("INVALID_ARGS", "token 为空", null)
      return
    }
    val code = engine.renewToken(token)
    result.success(code)
  }

  fun handleSetClientRole(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val role = (args["role"] as? Number)?.toInt()
    if (role == null) {
      result.error("INVALID_ARGS", "role 为空", null)
      return
    }
    val latencyLevel = (args["latencyLevel"] as? Number)?.toInt()
    if (latencyLevel == null) {
      val code = engine.setClientRole(role)
      result.success(code)
      return
    }
    val options = ClientRoleOptions()
    options.audienceLatencyLevel = when (latencyLevel) {
      1 -> Constants.AUDIENCE_LATENCY_LEVEL_LOW_LATENCY
      2 -> Constants.AUDIENCE_LATENCY_LEVEL_ULTRA_LOW_LATENCY
      else -> Constants.AUDIENCE_LATENCY_LEVEL_LOW_LATENCY
    }
    val code = engine.setClientRole(role, options)
    result.success(code)
  }

  fun handleMuteAllRemoteAudioStreams(call: MethodCall, result: Result) {
    mediaActions.handleMuteAllRemoteAudioStreams(call, result)
  }

  fun handleMuteAllRemoteVideoStreams(call: MethodCall, result: Result) {
    mediaActions.handleMuteAllRemoteVideoStreams(call, result)
  }

  fun handleMuteRemoteAudioStream(call: MethodCall, result: Result) {
    mediaActions.handleMuteRemoteAudioStream(call, result)
  }

  fun handleMuteRemoteVideoStream(call: MethodCall, result: Result) {
    mediaActions.handleMuteRemoteVideoStream(call, result)
  }

  fun handleMuteLocalAudioStream(call: MethodCall, result: Result) {
    mediaActions.handleMuteLocalAudioStream(call, result)
  }

  fun handleMuteLocalVideoStream(call: MethodCall, result: Result) {
    mediaActions.handleMuteLocalVideoStream(call, result)
  }

  fun handleSetRemoteVideoStreamType(call: MethodCall, result: Result) {
    mediaActions.handleSetRemoteVideoStreamType(call, result)
  }

  fun handleEnableVideo(call: MethodCall, result: Result) {
    mediaActions.handleEnableVideo(call, result)
  }

  fun handleEnableLocalVideo(call: MethodCall, result: Result) {
    mediaActions.handleEnableLocalVideo(call, result)
  }

  fun handleStartPreview(call: MethodCall, result: Result) {
    mediaActions.handleStartPreview(call, result)
  }

  fun handleStopPreview(call: MethodCall, result: Result) {
    mediaActions.handleStopPreview(call, result)
  }

  fun handleTakeSnapshot(call: MethodCall, result: Result) {
    mediaActions.handleTakeSnapshot(call, result)
  }

  fun handleStartRecording(call: MethodCall, result: Result) {
    recordingActions.handleStartRecording(call, result)
  }

  fun handleStopRecording(result: Result) {
    recordingActions.handleStopRecording(result)
  }

  private fun buildChannelMediaOptions(optionsMap: Map<*, *>?): ChannelMediaOptions {
    val options = ChannelMediaOptions()
    if (optionsMap == null) {
      return options
    }
    (optionsMap["publishCameraTrack"] as? Boolean)?.let { options.publishCameraTrack = it }
    (optionsMap["publishMicrophoneTrack"] as? Boolean)?.let { options.publishMicrophoneTrack = it }
    (optionsMap["autoSubscribeAudio"] as? Boolean)?.let { options.autoSubscribeAudio = it }
    (optionsMap["autoSubscribeVideo"] as? Boolean)?.let { options.autoSubscribeVideo = it }
    (optionsMap["audienceLatencyLevel"] as? Number)?.toInt()?.let { options.audienceLatencyLevel = it }
    (optionsMap["channelProfile"] as? Number)?.toInt()?.let { options.channelProfile = it }
    (optionsMap["clientRoleType"] as? Number)?.toInt()?.let { options.clientRoleType = it }
    return options
  }
}

File: ios/Classes/AgoraRtcEngineActions.swift
import Flutter
import AgoraRtcKit

final class AgoraRtcEngineActions {
  private let engineHolder: AgoraRtcEngineViewHolder
  private let eventHandler: AgoraRtcEngineEventHandler
  private let mediaActions: AgoraRtcEngineMediaActions
  private let recordingActions: AgoraRtcEngineRecordingActions

  init(engineHolder: AgoraRtcEngineViewHolder, emitter: AgoraEventEmitter) {
    self.engineHolder = engineHolder
    self.eventHandler = AgoraRtcEngineEventHandler(emitter: emitter)
    self.mediaActions = AgoraRtcEngineMediaActions(engineHolder: engineHolder)
    self.recordingActions = AgoraRtcEngineRecordingActions(engineHolder: engineHolder, emitter: emitter)
  }

  func handleCreateEngine(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    if engineHolder.rtcEngine != nil {
      result(FlutterError(code: "ENGINE_EXISTS", message: "RtcEngine 已初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let appId = args["appId"] as? String, !appId.isEmpty else {
      result(FlutterError(code: "INVALID_ARGS", message: "appId 为空", details: nil))
      return
    }
    let config = AgoraRtcEngineConfig()
    config.appId = appId
    let engine = AgoraRtcEngineKit.sharedEngine(with: config, delegate: eventHandler)
    engineHolder.rtcEngine = engine
    engineHolder.bindCachedVideoViews()
    recordingActions.clear()
    result(0)
  }

  func handleDestroyEngine(result: @escaping FlutterResult) {
    if engineHolder.rtcEngine != nil {
      AgoraRtcEngineKit.destroy()
      engineHolder.clear()
    }
    recordingActions.clear()
    result(0)
  }

  func handleJoinChannel(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    let token = (args["token"] as? String) ?? ""
    guard let channelId = args["channelId"] as? String, !channelId.isEmpty else {
      result(FlutterError(code: "INVALID_ARGS", message: "channelId 为空", details: nil))
      return
    }
    let uid = (args["uid"] as? NSNumber)?.uintValue ?? 0
    let optionsMap = args["options"] as? [String: Any]
    let options = buildChannelMediaOptions(optionsMap)
    let code = engine.joinChannel(byToken: token, channelId: channelId, uid: uid, mediaOptions: options)
    result(code)
  }

  func handleLeaveChannel(result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    let code = engine.leaveChannel(nil)
    result(code)
  }

  func handleUpdateChannelMediaOptions(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let optionsMap = args["options"] as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "options 为空", details: nil))
      return
    }
    let options = buildChannelMediaOptions(optionsMap)
    let code = engine.updateChannel(with: options)
    result(code)
  }

  func handleSetChannelProfile(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let profile = (args["profile"] as? NSNumber)?.intValue else {
      result(FlutterError(code: "INVALID_ARGS", message: "profile 为空", details: nil))
      return
    }
    guard let profileType = AgoraChannelProfile(rawValue: profile) else {
      result(FlutterError(code: "INVALID_ARGS", message: "profile 不合法", details: nil))
      return
    }
    let code = engine.setChannelProfile(profileType)
    result(code)
  }

  func handleRenewToken(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let token = args["token"] as? String, !token.isEmpty else {
      result(FlutterError(code: "INVALID_ARGS", message: "token 为空", details: nil))
      return
    }
    let code = engine.renewToken(token)
    result(code)
  }

  func handleSetClientRole(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let roleValue = (args["role"] as? NSNumber)?.intValue,
          let role = AgoraClientRole(rawValue: roleValue) else {
      result(FlutterError(code: "INVALID_ARGS", message: "role 为空", details: nil))
      return
    }
    let latencyLevel = (args["latencyLevel"] as? NSNumber)?.intValue
    if latencyLevel == nil {
      let code = engine.setClientRole(role)
      result(code)
      return
    }
    let options = AgoraClientRoleOptions()
    options.audienceLatencyLevel = latencyLevel == 2 ? .ultraLowLatency : .lowLatency
    let code = engine.setClientRole(role, options: options)
    result(code)
  }

  func handleMuteAllRemoteAudioStreams(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    mediaActions.handleMuteAllRemoteAudioStreams(call, result: result)
  }

  func handleMuteAllRemoteVideoStreams(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    mediaActions.handleMuteAllRemoteVideoStreams(call, result: result)
  }

  func handleMuteRemoteAudioStream(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    mediaActions.handleMuteRemoteAudioStream(call, result: result)
  }

  func handleMuteRemoteVideoStream(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    mediaActions.handleMuteRemoteVideoStream(call, result: result)
  }

  func handleMuteLocalAudioStream(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    mediaActions.handleMuteLocalAudioStream(call, result: result)
  }

  func handleMuteLocalVideoStream(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    mediaActions.handleMuteLocalVideoStream(call, result: result)
  }

  func handleSetRemoteVideoStreamType(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    mediaActions.handleSetRemoteVideoStreamType(call, result: result)
  }

  func handleEnableVideo(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    mediaActions.handleEnableVideo(call, result: result)
  }

  func handleEnableLocalVideo(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    mediaActions.handleEnableLocalVideo(call, result: result)
  }

  func handleStartPreview(result: @escaping FlutterResult) {
    mediaActions.handleStartPreview(result: result)
  }

  func handleStopPreview(result: @escaping FlutterResult) {
    mediaActions.handleStopPreview(result: result)
  }

  func handleTakeSnapshot(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    mediaActions.handleTakeSnapshot(call, result: result)
  }

  func handleStartRecording(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    recordingActions.handleStartRecording(call, result: result)
  }

  func handleStopRecording(result: @escaping FlutterResult) {
    recordingActions.handleStopRecording(result: result)
  }

  private func buildChannelMediaOptions(_ optionsMap: [String: Any]?) -> AgoraRtcChannelMediaOptions {
    let options = AgoraRtcChannelMediaOptions()
    guard let map = optionsMap else {
      return options
    }
    if let publishCameraTrack = map["publishCameraTrack"] as? Bool {
      options.publishCameraTrack = publishCameraTrack
    }
    if let publishMicrophoneTrack = map["publishMicrophoneTrack"] as? Bool {
      options.publishMicrophoneTrack = publishMicrophoneTrack
    }
    if let autoSubscribeAudio = map["autoSubscribeAudio"] as? Bool {
      options.autoSubscribeAudio = autoSubscribeAudio
    }
    if let autoSubscribeVideo = map["autoSubscribeVideo"] as? Bool {
      options.autoSubscribeVideo = autoSubscribeVideo
    }
    if let roleValue = (map["clientRoleType"] as? NSNumber)?.intValue,
       let role = AgoraClientRole(rawValue: roleValue) {
      options.clientRoleType = role
    }
    if let latencyLevel = (map["audienceLatencyLevel"] as? NSNumber)?.intValue {
      options.audienceLatencyLevel = latencyLevel == 2 ? .ultraLowLatency : .lowLatency
    }
    if let profileValue = (map["channelProfile"] as? NSNumber)?.intValue,
       let profile = AgoraChannelProfile(rawValue: profileValue) {
      options.channelProfile = profile
    }
    return options
  }

}
